### PR TITLE
Implements GET /mentor & /mentee + Auth Layer

### LINKED ISSUE #64: Add an auth layer for backend
**Requirements for feature**
- [ ] Defined such that it is easy to cover a route with the auth layer (all routes except signup will be covered)
- [ ] This auth layer will perform ALL authentication checks using the user token

**Any context for feature**
This is critical in moving our backend towards release because all routes needs to be authenticated (except signup). 

**Resources / tips for development**
- [ ] Look into how to effectively use middleware functions to setup this auth layer
- [ ] Benefit of having a middleware function with the auth logic is, we can add that middleware to the route definition whenever we need so very easy to authenticate a route.

### PR DESCRIPTION
**Pull request by adhi0331**
### Tracking Info

Resolves #64 , #37 

Make sure your branch name conforms to: `<feature|staging|bugfix|...>/<username>/<3-4 word description separated by dashes>`. Otherwise, please rename your branch and create a new PR.

### Changes

What changes did you make?

- Added the get mentor/mentee route
- Added a get image route
- Added an auth layer to verify user credentials 

Note that the auth layer is something that will be used in all routes going forward. The create user route was updated to add custom roles to each user. This is our way of checking whether the user making a request is a mentor or mentee. 

Here is how the get route works on a high level:

1. Verify that the token is valid via Firebase API
2. Update the req body to include user type (Mentee or Mentor
3. For get Mentee...
4. If it is a mentee calling the route give all information, if its a mentor give only limited necessary information (see figma)
5. For get Mentor
6. If it is a mentor calling the route give all information, if its a mentee calling the route give limited information (see figma)
7. To get the image use the get image route

Something to also note is that the default ALUM image is stored once on the database. This is to optimize storage space. When creating a new user the image id will be set to the id of the default image stored in the database. This id will also be stored as an env variable. 

### Testing

How did you confirm your changes worked? 

I ran the commands on postman. I got the auth token using the post command that was sent on slack. I also tested the various edge cases which included auth errors and service errors.

### Confirmation of Change 

Upload a screenshot, if possible. Otherwise, please provide instructions on how to see the change.

Get Image Route (for both mentor and mentee)

Get image

Mentee get Mentee

Mentor get Mentee

Mentee get Mentor

Mentee get mentor

Mentor get Mentor

Mentor get mentor

Mentor get Mentee
Mentor get Mentee

### CODE REVIEW

File: backend/src/routes/user.ts [hunk 1]
```diff
@@ -112,4 +120,160 @@ router.post("/mentor", [validateMentor], async (req: Request, res: Response) =>
...
 
+/**
+ * This is a get route for a mentee. Note that the response is dependant on
+ * the person calling the method
+ *
+ * Mentee calling: gain all info about the mentee
+ *
+ * Mentor calling: mentee name, image, grade, about, career interests, topics of interest
+ */
+router.get("/mentee/:userId", [verifyAuthToken, upload], async (req: Request, res: Response) => {
```
Comment (AmanKAggarwal) [inline backend/src/routes/user.ts:131]: Need to add `mentorId` to payload

File: backend/src/routes/user.ts [hunk 2]
```diff
@@ -112,4 +120,160 @@ router.post("/mentor", [validateMentor], async (req: Request, res: Response) =>
...
+/**
+ * This is a get route for a mentor. Note that the response is dependant on the person
+ * calling the method
+ *
+ * Mentee calling: mentor name, image, major, minor, college, career, graduation year, calendlyLink,
+ * why were you paired?, about, topics of experties
+ *
+ * Mentor calling: gain all info about the mentor
+ */
+router.get("/mentor/:userId", [verifyAuthToken, upload], async (req: Request, res: Response) => {
```
Comment (AmanKAggarwal) [inline backend/src/routes/user.ts:202]: Need to add 
- menteeId (if request is sent by the mentor's mentee)
- menteeIds (if requested by the mentor themselves)

File: backend/src/config.ts [hunk 3]
```diff
@@ -13,34 +14,42 @@ dotenv.config({
...
 let serviceAccountKeyV = "";
+let defaultImageIdV = "";
 
 /**
  * Todo: these should throw errors instead of logging messages
  */
 if (!process.env.APP_PORT) {
-  console.log("Could not find app port env variable");
+  throw InternalError.NO_APP_PORT;
 } else {
```
Comment (wllmwu) [inline backend/src/config.ts:24]: Nice

**Review by wllmwu — COMMENTED** [review]
LGTM

File: backend/src/routes/user.ts (hunk 2 above)
Comment (adhi0331) [reply backend/src/routes/user.ts:207]: Fixed!

File: backend/src/routes/user.ts (hunk 1 above)
Comment (adhi0331) [reply backend/src/routes/user.ts:136]: Fixed!

File: backend/src/errors/auth.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+/**
```
Comment (AmanKAggarwal) [inline backend/src/errors/auth.ts:1]: Great! Love having errors in one place

File: backend/src/errors/auth.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+import { CustomError } from "./errors";
+
+const DECODE_ERROR = "Error decoding the auth token. Make sure the auth token is valid";
+const TOKEN_NOT_IN_HEADER =
+  "Token was not found in the header. Be sure to use Bearer <Token> syntax";
+const INVALID_AUTH_TOKEN = "Token was invalid. Be sure to refresh token if needed.";
+
+export class AuthError extends CustomError {
+  static DECODE_ERROR = new AuthError(0, 401, DECODE_ERROR);
```
Comment (AmanKAggarwal) [inline backend/src/errors/auth.ts:13]: Can you also update these error status on the API Sheet?
https://docs.google.com/spreadsheets/d/1h1LmmF9NtEkucmPMpy8WkK-ubjwUXb_qWc0b0USPN04/edit#gid=0

File: backend/src/errors/internal.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+const ERROR_GETTING_MENTEE = "There was some error getting the mentee";
+const ERROR_GETTING_MENTOR = "There was some error getting the mentor";
+const ERROR_GETTING_IMAGE = "There was some error getting the image";
+const NO_APP_PORT = "Could not find app port env variable";
+const NO_MONGO_URI = "Could not find mongo uri env variable";
+const NO_SERVICE_ACCOUNT_KEY = "Could not find service account key env variable";
+const NO_DEFAULT_IMAGE_ID = "Could not find default image id env variable";
+
+export class InternalError extends CustomError {
+  static ERROR_GETTING_MENTEE = new InternalError(0, 403, ERROR_GETTING_MENTEE);
```
Comment (AmanKAggarwal) [inline backend/src/errors/internal.ts:16]: These as well. Just the status, a brief message about the status is good. 
https://docs.google.com/spreadsheets/d/1h1LmmF9NtEkucmPMpy8WkK-ubjwUXb_qWc0b0USPN04/edit#gid=0

File: backend/src/middleware/auth.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
+/**
```
Comment (AmanKAggarwal) [inline backend/src/middleware/auth.ts:1]: great stuff! This layer will be super useful

File: backend/src/routes/user.ts [hunk 8]
```diff
@@ -35,9 +44,10 @@ router.post("/mentee", [validateMentee], async (req: Request, res: Response) =>
       topicsOfInterest,
       careerInterests,
       mentorshipGoal,
+      mentorID,
       status,
     });
-    await createUser(mentee._id.toString(), email, password);
+    await createUser(mentee._id.toString(), email, password, "mentee");
```
Comment (AmanKAggarwal) [inline backend/src/routes/user.ts:50]: nice researching!!

File: backend/src/routes/user.ts [hunk 9]
```diff
@@ -112,4 +125,160 @@ router.post("/mentor", [validateMentor], async (req: Request, res: Response) =>
...
 
+/**
+ * This is a get route for a mentee. Note that the response is dependant on
+ * the person calling the method
+ *
+ * Mentee calling: gain all info about the mentee
+ *
+ * Mentor calling: mentee name, image, grade, about, career interests, topics of interest
+ */
+router.get("/mentee/:userId", [verifyAuthToken, upload], async (req: Request, res: Response) => {
```
Comment (AmanKAggarwal) [inline backend/src/routes/user.ts:136]: Why do we need upload for a GET request? Isn't `upload` just for the case where we upload an image into the server?

**Review by AmanKAggarwal — APPROVED** [review]
Just had one question and few notes but other than that looks great! Good stuff

File: backend/src/errors/service.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+const IMAGE_NOT_SAVED =
+  "Image was not able to be saved. Be sure to specify that image key is image";
+const IMAGE_NOT_FOUND = "Image was not found. Please make sure id passed in route is valid";
+const INVALID_MONGO_ID = "Mongo Id was invalid. Please ensure that the id is correct";
+const MENTEE_WAS_NOT_FOUND = "Mentee was not found";
+const MENTOR_WAS_NOT_FOUND = "Menor was not found";
+const IMAGE_WAS_NOT_FOUND = "Image was not found";
+
+export class ServiceError extends CustomError {
+  static IMAGE_NOT_SAVED = new ServiceError(0, 402, IMAGE_NOT_SAVED);
```
Comment (AmanKAggarwal) [inline backend/src/errors/service.ts:17]: HTTP 402 code means "Payment Required" which does not go well here. 

You could use [404 - Not Found](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.4.5) for these cases

File: backend/src/errors/internal.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+const ERROR_GETTING_MENTEE = "There was some error getting the mentee";
+const ERROR_GETTING_MENTOR = "There was some error getting the mentor";
+const ERROR_GETTING_IMAGE = "There was some error getting the image";
+const NO_APP_PORT = "Could not find app port env variable";
+const NO_MONGO_URI = "Could not find mongo uri env variable";
+const NO_SERVICE_ACCOUNT_KEY = "Could not find service account key env variable";
+const NO_DEFAULT_IMAGE_ID = "Could not find default image id env variable";
+
+export class InternalError extends CustomError {
```
Comment (AmanKAggarwal) [inline backend/src/errors/internal.ts:15]: Let's use 500 - Internal Server Error for these. 
403 stands for Forbidden which does not apply here

File: backend/src/routes/image.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+      .status(ServiceError.INVALID_MONGO_ID.status)
+      .send(ServiceError.INVALID_MONGO_ID.message);
+  }
+
+  try {
+    const image = await Image.findById(imageId);
+    if (!image) {
+      throw ServiceError.IMAGE_NOT_FOUND;
+    }
+    return res.status(201).set("Content-type", image.mimetype).send(image.buffer);
```
Comment (AmanKAggarwal) [inline backend/src/routes/image.ts:32]: Error code 201 stands for Creation Success. For GET, we should use 200 OK error code for success.

**Review by christen03 — APPROVED** [review]
looks good! great job

File: backend/src/routes/user.ts (hunk 9 above)
Comment (adhi0331) [reply backend/src/routes/user.ts]: It was there to test uploading an image to MongoDB but it isn't needed for the get route.
